Documentation/contributors/design-proposals/operator-validation-library.md

```diff
@@ -0,0 +1,55 @@
+
+Each of these validates only a piece of the bundle and most are not in sync with changes to the CSV type. This means operator owners use different tools to validate their bundle - each providing partial or inconsistent validation - which dampens the user experience. The operator framework needs to agree on a single source of truth for a valid OLM-managed operator.
+
+Our plan is to bring the already defined validators scattered across various code bases (along with potentially new validators) into a single location/package with validation logic that adapts to changes made to OLM.
+
+### Stakeholders
+
+Our main stakeholders of this library are:
+
+- OLM/operator-registry: refactoring some code to call a single validation library may help with improving user experience.
+- Operator-sdk: calling this library for the csv-gen tool instead of creating custom functions.
+- Operator-courier (or similar tool): community-operators or upstream operator owners who are not using the sdk to build their operator need a validation tool.
+- OSD SRE: validating bundles before pushing them down their deployment pipeline to minimize resource expenditure.
+
+### Development Strategy
+
+As a preliminary step, this library is interested in static verification of operator bundles. Static verification is crucial to help improve the development process of operators. It is time efficient and can be used in a variety of contexts: from IDE linter extensions to CI pipelines with limited resources to command line utilities. All consistent by calling the same APIs from this library.
+
+In the future, we aim to extend this library to contain runtime tests and apis to be called by runtime tools on cluster.
+
+### Requirements
+
+- Methods of this library should adapt to any changes in OLM&#39;s type. There should be minimal to no changes required when there is a change in OLM&#39;s type and/or version. Or if this is not possible, methods of the library need to be part of the build/test pipeline of OLM.
+- Better error reporting by informing the user of all errors found instead of exiting after the first occurrence. We should define custom error types easily digestible by users of these APIs.
+- APIs should expose verification methods for individual yaml files as well as for the bundle as a whole.
+- This library should expose a validator interface for users to define their own custom validators (ex: operatorhub.io specific validators) while obeying the same overarching structure.
```

Comment:
How will we take advantage in practice? Have an operatorhub.io version of the validator tool? or can you load in additional validators at runtime?

Reply:
What we were thinking was that operatorhub.io folks would write a validator using the interface exposed in this library. Then whatever command line tool (that replaces what courier is doing) would call the apis from the library **and** load those validators defined by operatorhub.io (wherever those live). That tool could easily digest any number of such validators since they would have the same interface.